### PR TITLE
Make bib parser configurable

### PR DESCRIPTION
**Pull request by bdarcus**
Implement #680.

This should be enough to enable #397?

Please review and comment. 

--------------

## On the org-bibtex parser, etc

I don't think I have time to do the org-bibtex parser and updatep functions. Would be great if someone can test this with that.

The first should look something like the below, and return a hash-table exactly like `parsebib-parse` does:

```elisp
(cl-defun citar-cache--org-bibtex-parser (files &optional &key entries fields)
  "Return a parsebib-compatible hash-table for a list of org-bibtex FILES."
  ;; FILES is a list of file paths, though in citar, will only be a single item list
  (map-into                                ; convert alist into hash-table
    (org-element-cache-map ...
```

EDIT - I tried to figure out how to add the parser to parsebib on this branch:

https://github.com/bdarcus/parsebib/tree/org-bibtex

But I ran into a UX and technical issue: to take advantage of `org-element-cache-map`, you can't use temp-buffers, and need to be in the org major-mode. So unlike the other formats, it would seem to be more sensible to open all listed org files. But that's a departure from how parsebib and citar work. And what happens if someone does something crazy like do file-per-reference using org?

This may suggest that parser is better off here, and we just clearly document the different workflows.

### CODE REVIEW

File: citar-cache.el [hunk 1]
```diff
@@ -218,6 +219,19 @@ parsed is different."
             (equal (plist-get oldprops :hash) (plist-get newprops :hash))
             (equal (plist-get oldprops :fields) (plist-get newprops :fields)))))
 
+;; Parsers
+
+(defvar citar-cache--parser
+  (list (cons t #'parsebib-parse)))
```
Comment (bdarcus) [inline citar-cache.el:225]: This maybe should be a triplet, with the last arg an updatep function.

File: citar-cache.el [hunk 2]
```diff
@@ -209,7 +211,7 @@ re-hashing the file contents."
                    (buffer-hash)))))
     `(:size ,size :mtime ,mtime :hash ,hash :fields ,fields)))
 
-(defun citar-cache--update-bibliography-p (oldprops newprops)
+(defun citar-cache--update-parsebib-bibliography-p (oldprops newprops)
```
Comment (bdarcus) [inline citar-cache.el:214]: This probably needs a more general name. Maybe "parsebib" -> "default"?

File: citar-cache.el [hunk 3]
```diff
@@ -209,7 +211,7 @@ re-hashing the file contents."
                    (buffer-hash)))))
     `(:size ,size :mtime ,mtime :hash ,hash :fields ,fields)))
 
-(defun citar-cache--update-bibliography-p (oldprops newprops)
+(defun citar-cache--update-default-bibliography-p (oldprops newprops)
```
Comment (bdarcus) [inline citar-cache.el:214]: One question on this; what's the implications of the very last line of this docstring for this PR?

File: citar-cache.el [hunk 4]
```diff
@@ -218,6 +220,38 @@ parsed is different."
             (equal (plist-get oldprops :hash) (plist-get newprops :hash))
             (equal (plist-get oldprops :fields) (plist-get newprops :fields)))))
 
+;; Parsers
+
+(defconst citar-cache--parser
+  (list
+   (cons t (cons #'parsebib-parse #'citar-cache--update-default-bibliography-p)))
```
Comment (bdarcus) [inline citar-cache.el:227]: This is maybe a bit ugly; does that matter?

**Comment by aikrahguzar** [comment]
I am not sure how much this will help. The configuration it introduces is seems unlikely to be used outside of citar since a new parser is likely a lot of work. If the intention is to allow formats not supported by parsebib think a simpler way would be to essentially define an analog of `parsebib-parse` that does the calls to `parsebib-parse-bib-buffer` etc as `parsebib-parse` and includes the additional calls too. If a user really wants to support some format outside of citar, this function can be advised.

**Comment by bdarcus** [comment]
That's a good point.

I went this way because I wasn't sure if we want the org-bibtex parser in the citar package, or in a separate package. I'm still not sure on that actually.

WDYT?

**Comment by aikrahguzar** [comment]
> I went this way because I wasn't sure if we want the org-bibtex parser in the citar package, or in a separate package. I'm still not sure on that actually, largely because there are outstanding UX and performance issues to sort out there.
> 
> WDYT?

So the problematic scenario is when the parser is in a separate package and we don't want to depend on it? I think the soft dependency can be handled via `featurep` check inside the citar version of  `parsebib-parse` and in that case we can warn users that they need the package installed if we are called to parse an org-bibtex file.

**Comment by bdarcus** [comment]
@aikrahguzar 

> So the problematic scenario is when the parser is in a separate package and we don't want to depend on it?

That's the immediate issue I was thinking of. But perhaps there could be other in the future?

**Pull request closed without merging**